pkg/controller/infra/infra_controller.go

```diff
@@ -0,0 +1,240 @@
+	informers "github.com/openshift/cluster-operator/pkg/client/informers_generated/externalversions/clusteroperator/v1alpha1"
+	lister "github.com/openshift/cluster-operator/pkg/client/listers_generated/clusteroperator/v1alpha1"
+	"github.com/openshift/cluster-operator/pkg/controller"
+)
+
+const (
+	// maxRetries is the number of times a service will be retried before it is dropped out of the queue.
+	// With the current rate-limiter in use (5ms*2^(maxRetries-1)) the following numbers represent the
+	// sequence of delays between successive queuings of a service.
+	//
+	// 5ms, 10ms, 20ms, 40ms, 80ms, 160ms, 320ms, 640ms, 1.3s, 2.6s, 5.1s, 10.2s, 20.4s, 41s, 82s
+	maxRetries = 15
+
+	controllerLogName = "infra"
+)
+
+// NewInfraController returns a new *InfraController.
+func NewInfraController(clusterInformer informers.ClusterInformer, kubeClient kubeclientset.Interface, clusteroperatorClient clusteroperatorclientset.Interface) *InfraController {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	// TODO: remove the wrapper when every clients have moved to use the clientset.
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events("")})
+
+	if kubeClient != nil && kubeClient.CoreV1().RESTClient().GetRateLimiter() != nil {
+		metrics.RegisterMetricAndTrackRateLimiterUsage("clusteroperator_cluster_controller", kubeClient.CoreV1().RESTClient().GetRateLimiter())
+	}
+
+	logger := log.WithField("controller", controllerLogName)
+	c := &InfraController{
+		coClient: clusteroperatorClient,
+		queue:    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "cluster"),
+		logger:   logger,
+	}
+
+	clusterInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.addCluster,
+	})
+	c.clustersLister = clusterInformer.Lister()
+	c.clustersSynced = clusterInformer.Informer().HasSynced
+
+	c.syncHandler = c.syncCluster
+	c.enqueueCluster = c.enqueue
+
+	return c
+}
+
+// InfraController manages clusters.
+type InfraController struct {
+	coClient clusteroperatorclientset.Interface
+
+	// To allow injection of syncCluster for testing.
+	syncHandler func(hKey string) error
+
+	// used for unit testing
+	enqueueCluster func(cluster *clusteroperator.Cluster)
+
+	// clustersLister is able to list/get clusters and is populated by the shared informer passed to
+	// NewClusterController.
+	clustersLister lister.ClusterLister
+	// clustersSynced returns true if the cluster shared informer has been synced at least once.
+	// Added as a member to the struct to allow injection for testing.
+	clustersSynced cache.InformerSynced
+
+	// Clusters that need to be synced
+	queue workqueue.RateLimitingInterface
+
+	logger *log.Entry
+}
+
+func (c *InfraController) addCluster(obj interface{}) {
+	cluster := obj.(*clusteroperator.Cluster)
+	c.logger.Debugf("adding cluster %s", cluster.Name)
```

Comment:
Its unfortunate that logrus does not have more granular logging levels than the 6.

Reply:
I could see the argument for one more "insane" level below debug but beyond that I think it gets a bit confusing.

I'm also not crazy about the mixing or loss of glog statements from the kube infra we're re-using. 

However for me the pros outweigh the cons. Our logs are what I'm interested in and we should be able to assume the kube infra we're re-using works, at least at some point once it's wired up. If we do end up with controllers in their own pods then we no longer will care much about glog output from the kube infra.

debug would be our highly verbose mode purely for debugging. (possibly re-use that to automatically re-enable glog once we shut it off), info is more high level and informative but still suitable to run in prod, and warn if you only care that something is, or might be wrong. I would envision that glog is silent by default once we reach a certain point.

The structured logging approach *should*  be really beneficial to us if we have the ELK stack in our final cluster, and from what I can tell a lot of projects in the kube/docker ecosystem are pushing away from glog to logrus. This should help us think more intelligently about how we'll log, and query those logs in prod.

I can be swayed though, WDYT?

Reply:
The glogs coming from kube aren't only useful to trace errors in the infrastructure. They are also useful to trace errors in how our code is interacting with the infrastructure. For example, if a controller is not sending the correct information in a status update for a resource, that can be seen in the logs coming from the infrastructure.

It may very well be that the formalized structure that logrus gives in its log output obviates the usefulness of more granular log levels. I am willing to be amazed.